README.md
# deco
R package
---------------------------------------------------
DECO:
Decomposing heterogeneous population cohorts for patient stratification and discovery of biomarkers using omic data profiling.
---------------------------------------------------
We present DECO, a new bioinformatic method to explore and find differences in heterogeneous large datasets usually produced in biological or biomedical omic-wide studies. This method makes a comprehensive analysis of multidimensional datasets, consisting on a collection of samples where hundreds or thousands of features have been measured with a large-scale high-throughput technology, for example, a genomic or proteomic technique. The method finds the differences in the profiles of the features along the samples and identifies the associations between them, showing the features that best mark a given class or category as well as possible sample outliers that do not follow the same pattern of the majority of the corresponding cohort. Interestingly, it can be used for the comparison of two or more classes of samples or for unsupervised comparisons. **DECO** allows the discovery of multiple classes or categories and is quite adequate for patients stratification. 

The statistical procedure followed in both parts of the method are detailed in the original [publication](#references) [1]. A detailed **vignette** is included to explain how to use **DECO** for the analysis of multidimensional datasets, which may include heterogeneous samples or categories. The aim is to improve characterization and stratification of complex sample series, mostly focusing on large patient cohorts, where the existence of outlier or mislabeled samples is quite possible.

**DECO** performs a recursive exploration of differential signal changes between samples, finding variables assigned to: 
**(i)** the main classes or groups of samples that are in the studied cohorts 
**(ii)** significant variation or alteration among certain individuals (related or not to an a-priori known class) 
**(iii)** outlier patterns within feature profiles 
**(iv)** sample outliers (i.e. individuals that behave in a different way to the main groups and have specific markers). 

![Workflow](https://github.com/fjcamlab/deco/blob/master/vignettes/figures/flow.jpg)

---------------------------------------------------
**INSTALLATION**

The **deco** R source package can be directly downloaded from **[Bioconductor repository](https://bioconductor.org/packages/deco/)** or **[GitHub repository](https://github.com/fjcamlab/deco)**. This R package contains a experimental dataset as example, two pre-run R objects and all functions needed to run a DECO analysis.

```r
## Bioconductor repository
if (!requireNamespace("BiocManager", quietly = TRUE)) {
      install.packages("BiocManager")
  }
BiocManager::install("deco")

## GitHub repository using devtools
BiocManager::install("devtools")
devtools::install_github("fjcamlab/deco")
```

---------------------------------------------------
**EXAMPLE OF DIFFERENTIAL ANALYSIS USING DECO**

**DECO** R package has been designed to perform a deep analysis of heterogeneous populations through two main steps: (i) RDA, including a subsampling procedure (`decoRDA()`) based on LIMMA and (ii) a posterior integration with NSCA (`decoNSCA()`) to find out new subclasses of samples. As a consequence of these two steps, **DECO** will calculate a new *h-statistic* which replaces the original omic data to improve the sample stratification. The *h-statistic* integrates both omic dispersion and predictor-response information given by NSCA.

Pipeline using two categories of samples to compare:
```r
### Loading the R packages
library(deco)
library(BiocParallel) # for parallel computation

# Computing in shared memory
bpparam <- MulticoreParam()

########################
# Loading example data #
########################
## Data from two subtypes (ALK+ and ALK-) of Anaplastic Large Cell Leukemia (ALCL).
data(ALCLdata)

# It includes a subset from an ALCL transcriptomic dataset (GEO id: GSE65823).
# to see the example SummarizedExperiment object
ALCL

## Classes vector to run a binary analysis to compare both classes.
classes.ALCL <- colData(ALCL)[,"Alk.positivity"]
names(classes.ALCL) <- colnames(ALCL)

#######################################################################
# RUNNING SUBSAMPLING OF DATA: BINARY design (two classes of samples) #
#######################################################################
# if annotation and rm.xy == TRUE, then
library(Homo.sapiens)

sub.ma.3r.1K <- decoRDA(data = assay(ALCL), classes = classes.ALCL, q.val = 0.01,
                rm.xy = TRUE, r = NULL, control = "pos", annot = FALSE, bpparam = bpparam,
                id.type = "ENSEMBL", iterations = 10000, pack.db = "Homo.sapiens")

#########################################################################################
# RUNNING NSCA STEP: Looking for subclasses within a category/class of samples compared #
#########################################################################################

deco.results.ma <- decoNSCA(sub = sub.ma.3r.1K, v = 80, method = "ward.D", bpparam = bpparam,
                       k.control = 3, k.case = 3, samp.perc = 0.05, rep.thr = 10)

# Phenotypical data from TCGA RNAseq samples.
colData(ALCL)

# h-statistic matrix used to stratify samples
hMatrix <- NSCAcluster(deco.results.ma)$Control$NSCA$h   #for control samples
hMatrix <- NSCAcluster(deco.results.ma)$Case$NSCA$h      #for case samples

```

Finally, the third main function `decoReport()` will generate a PDF file containing a summary of the analysis including: top-discriminant features, new subclasses of samples found, and several plots showing any relevant result of the analysis.

```r
########################################################
# PDF report with feature-sample patterns or subgroups #
########################################################
## Generate PDF report with relevant information and several plots.

## Binary example (ALK+ vs ALK-)
decoReport(deco.results.ma, sub.ma.3r.1K,
          pdf.file = "report_example_microarray_binary.pdf",
          info.sample = as.data.frame(colData(ALCL)[,8:10]),
          cex.names = 0.3, print.annot = TRUE)
```

---------------------------------------------------
**REFERENCES**

1: Campos-Laborie FJ, Risueño A, Roson-Burgo B, Droste C, Fontanillo C, Ortiz-Estevez M, Trotter MW, Sánchez-Santos JM and De Las Rivas J (2019). **DECO: decompose heterogeneous population cohorts for patient stratification and discovery of sample biomarkers using omic data profiling.** *Bioinformatics*. btz148, [doi.org/10.1093/bioinformatics/btz148](https://doi.org/10.1093/bioinformatics/btz148).

2: Ritchie ME, Phipson B, Wu D, Hu Y, Law CW, Shi W and Smyth GK (2015). **limma powers differential expression analyses for RNA-sequencing and microarray studies.** *Nucleic Acids Res.*, 43:e47. [doi:10.1093/nar/gkv007](https://doi:10.1093/nar/gkv007).



---------------------------------------------------
**NEWS**

Changes in version 0.99.46 (21-02-2019)
+ Bug fixes.
+ Annotation adapted to new "OrganismDbi" related packages. 
+ Three new diagnostic (plot) functions.
+ Enlarged vignette.

Changes in version 0.99.42 (17-12-2018)
+ Bug fixes.
+ Vignette converted into HTML format.
+ Accepted in Bioconductor. 

Changes in version 0.99.0 (06-11-2018)
+ Submitted to Bioconductor.



NEWS
Changes in version 1.0.1 (10-05-2019)
+ Release 3.9 Bioconductor
+ Changes in vignette.

Changes in version 0.99.53 (22-03-2019)
+ Bug fixes.

Changes in version 0.99.47 (21-02-2019)
+ Bug fixes.
+ Annotation adapted to new "OrganismDbi" related packages. 
+ Three new diagnostic (plot) functions.
+ Enlarged vignette.

Changes in version 0.99.42 (17-12-2018)
+ Bug fixes.
+ Vignette converted into HTML format.
+ Accepted in Bioconductor. 

Changes in version 0.99.0 (06-11-2018)
+ Submitted to Bioconductor.

R/plotGainingH.R
plotGainingH <- function(deco, data, ids, print.annot = FALSE,
                         orig.classes = TRUE) {
  
    if (all(!ids %in% rownames(deco@featureTable))) {
        stop("ERROR: Any provided IDs were found as significant by DECO.")
    }

    if (any(ids %in% rownames(deco@featureTable))) {
        msg1 <- c("NOTE: Following IDs have not been found by DECO, so they 
              will be not plotted:")
        msg2 <- paste(" ", ids[!(ids %in% rownames(deco@featureTable))], "\n ")
    }

    message(paste(msg1, msg2, sep = "\n"))

    if (print.annot & "SYMBOL" %in% colnames(featureTable(deco))) {
        msg <- "NOTE: IDs have been mapped to HGNC symbol."
        name <- deco@featureTable[ids[ids %in% rownames(deco@featureTable)], "SYMBOL"]
        names(name) <- ids[ids %in% rownames(deco@featureTable)]
    } else {
        msg <- "NOTE: Original feature IDs will be used."
        name <- ids[ids %in% rownames(deco@featureTable)]
        names(name) <- name
    }

    message(msg)

    ## Type of analysis run
    analysis <- .analysisType(deco)

    if (analysis == "Unsupervised" & orig.classes) {
        orig.classes <- FALSE
        message("NOTE: Unsupervised analysis is not associated to original classes,
            DECO subclasses will be represented.")
    }

    if (orig.classes) {
        classes <- deco@classes
    } else {
        if (analysis == "Binary") {
            classes <- c(
                deco@NSCAcluster[[1]]$samplesSubclass[, 1],
                deco@NSCAcluster[[2]]$samplesSubclass[, 1]
            )
        } else {
            classes <- deco@NSCAcluster[[1]]$samplesSubclass[, 1]
        }
    }

    ## Color
    color <- jColor(cbind(sort(classes)))

    ## Plotting
    for (id in ids[ids %in% rownames(deco@featureTable)]) {
        if (analysis == "Binary") {
            y <- c(
                deco@NSCAcluster[[1]]$NSCA$h[id, ],
                deco@NSCAcluster[[2]]$NSCA$h[id, ]
            )
        } else {
            y <- deco@NSCAcluster[[1]]$NSCA$h[id, ]
        }

        x <- data[id, names(y)]

        d <- data.frame(x, y)
        colnames(d) <- c("omic.data", "h.statistic")

        plot1 <- ggplot(
            data.frame(d, classes = classes[rownames(d)]),
            aes(y = omic.data, x = h.statistic, color = classes)
        ) + geom_point() +
            ggtitle(paste(name[id], "raw statistics")) +
            xlab("h-statistic") + ylab("omic data") +
            geom_smooth(method = "loess", span = 1, se = FALSE) + theme_minimal() +
            scale_color_manual(values = adjustcolor(names(color$ty), 0.7))

        d2 <- data.frame(apply(d, 2, function(x)
            rank(x, ties.method = "random")), classes = classes[rownames(d)])

        plot2 <- ggplot(d2, aes(y = omic.data, x = h.statistic, color = classes)) +
            geom_point() + ggtitle(paste(name[id], "rankings")) +
            xlab("ranking h-statistic") + ylab("ranking omic data") +
            geom_smooth(method = "loess", span = 1, se = FALSE) + theme_minimal() +
            scale_color_manual(values = adjustcolor(names(color$ty), 0.7))

        abc <- by(d2[, seq_len(2)], classes[rownames(d2)], data.frame)
        df <- lapply(names(abc), function(x) cbind(name = x, abc[[x]]))
        df <- reshape2::melt(do.call(rbind, df), id = "name")
        colnames(df) <- c("classes", "stat", "ranking")

        plot3 <- ggplot(df, aes(x = stat, y = ranking, color = classes)) +
            geom_boxplot(outlier.alpha = 0.6, col = rep(names(color$ty), each = 2), lwd = 0.5) +
            facet_grid(~classes) + theme_minimal()

        suppressWarnings(gridExtra::grid.arrange(plot1, plot2, plot3,
            layout_matrix = matrix(c(1, 2, 3, 3), byrow = TRUE, nrow = 2))
        )
    }
}


R/deco.R
#### DECO ####
## Algorithm to find differential significant markers in
## heterogeneous cohorts including a subsampling approach and posterior
## integration with a Non-Symmetrical Correspondence Analysis to create the
## new heterogeneity statistic (h-statistic).  Authors: Francisco J.
## Campos-Laborie, Jose Manuel Sanchez-Santos & Javier De Las Rivas
## Bioinformatics & Functional Genomics Group Cancer Research Center
## (CiC-IBMCC, CSIC/USAL) Salamanca, Spain http://www.cicancer.org/
## http://bioinfow.dep.usal.es

################################### decoRDA ### Iterative function to discover differentially expressed
################################### features using LIMMA R package (eBayes method) along the data.

decoRDA <- function(data, classes = NA, control = NA,
    r = NULL, q.val = 0.01, iterations = 10000, bpparam = SerialParam(), 
    annot = FALSE, id.type = NA, attributes = NA, 
    rm.xy = FALSE, pack.db = "Homo.sapiens") {
    call <- match.call()
    
    # Assessing 'data' input.
    msg <- NA
    if (is(data, "SummarizedExperiment") | is(data, "MultiAssayExperiment")) {
        if (!exists("SummarizedExperiment")) {
            msg <- c("ERROR: 'SummarizedExperiment' is not in library.")
        } else {
            requireNamespace("SummarizedExperiment", quietly = TRUE)
        }
        if (is(assay(data), "matrix")) {
            data <- assay(data)
        } else {
            msg <- c("ERROR: 'data' object must include only one 'assay'.")
        }
    }
    if (!is.matrix(data)) {
        msg <- c("Data must be a matrix or SummarizedExperiment object")
    }
    data <- as.matrix(data)
    
    if(any(duplicated(rownames(data))))
      stop("ERROR: Duplicated rownames are not allowed.")
    
    if (!is.na(msg)) {
        stop(msg)
    }
    
    # Setting up temporary directory to write intermediate results.
    temp.path <- .createTempFile()
    message("'temp' folder will be created to store internal loop data.")
    
    # Annotating IDs to chromosome location (locus) in order to remove those
    # placed on X or Y.  It is proposed to avoid significant results from
    # unbalanced gender contrasts.
    if (rm.xy & annot) {
        if (!exists(pack.db)) {
            stop("ERROR: Annotation package '", pack.db, "' has not been loaded.")
        } else {
            requireNamespace(pack.db, quietly = TRUE)
        }
        if (id.type %in% columns(x = get(pack.db))) {
            infogenes <- AnnotateDECO(ids = rownames(data), id.type = id.type, 
                attributes = c("TXCHROM", "TXCHROM"), pack.db = pack.db, verbose = FALSE)
            chrTest <- infogenes[infogenes[, "TXCHROM"] %in% c("X", "Y", "chrX", "chrY"), id.type]
            if (length(chrTest) > 1) {
                data <- data[rownames(infogenes)[!rownames(infogenes) %in% chrTest], ]
                msg <- paste(.timestamp(), "-- Features located in X or Y chromosome have been filtered:\n", 
                  length(chrTest), "features have been discarded.")
            } else {
                msg <- paste(.timestamp(), " All features are placed on X or Y chromosome,
                       filter will not be applied.")
            }
        } else {
            msg <- paste(.timestamp(), "It was not possible to assign chromosome location
                   to input IDs.")
        }
        message(msg)
    }

    ## Initial matrix for resampling design...
    results <- matrix(0, nrow = 1, ncol = 6)
    
    ## Defining classes and sample sizes
    unsup <- all(is.na(classes))
    multi <- FALSE
    if (!unsup) {
        classes <- factor(classes)
        if(length(levels(classes)) == 2 & is.na(control))
          control <- levels(classes)[1]
        
        combin <- .combCalcM(data, classes, control, 
                             iterations, multi, r, 
                             results)
        
        categories.control <- combin$categories.control
        categories.case <- combin$categories.case
        multi <- combin$multi
        cl1 <- combin$cl1
        cl2 <- combin$cl2
        n1 <- combin$n1
        n2 <- combin$n2
        results <- combin$results
        data <- combin$data
        iterations <- combin$iterations
    } else {
        n1 <- dim(data)[2]
        n2 <- dim(data)[2]
        categories.control <- seq_len(dim(data)[2])
        categories.case <- seq_len(dim(data)[2])
        message("\n Classes vector not defined as input.
                UNSUPERVISED analysis will be carry out.\n")
    }
    
    ## Setting up subsampling size 'r'.
    if ((is.null(r) || r <= 0)) {
        r <- round(sqrt(min(c(n1, n2))), digits = 0)
    }
    if (r < 3) {
        r <- 3
        message(" Resampling size set to 3.\n")
    }
    if (r > n1 || r > n2) {
        stop("ERROR: Resampling size can not be
             higher than samples.")
    }
    
    ## Message to user
    if (unsup) {
        msg <- paste(.timestamp(), "-- Resampling design:\n Unsupervised analysis for", 
            dim(data)[2], "samples.\n Resampling size:", r, "\n adj.p.value threshold:", 
            q.val)
    } else if (!multi) {
        msg <- paste(.timestamp(), "-- Resampling design:\n Binary analysis for", 
            dim(data)[2], "samples.\n Control or 0 --> '", cl1, "' with", 
            n1, "samples.\n Case or 1    --> '", cl2, "' with", n2, "samples.\n", 
            "Resampling size for both classes:", r, "\n adj.p.value threshold:", 
            q.val)
    } else {
        msg <- paste(.timestamp(), "-- Resampling design:\n Multiple supervised analysis for", 
            dim(data)[2], "samples and", length(levels(classes)), "classes:\n", 
            paste(levels(classes), collapse = "/"), ".\n Resampling size for all classes:", 
            r, "\n adj.p.value threshold:", q.val)
    }
    
    message(msg)
    
    ## Combinatorial calculations for binary and unsupervised designs
    results <- .combCalc(iterations, multi, classes, r, 
                         results, categories.control, 
                         categories.case, n1, n2)
    
    ## SUBSAMPLING STEP using LIMMA
    limmaRes <- .LIMMAcalc(data, results, classes, control, q.val, call, r, 
        temp.path, bpparam, multi, n1, n2)
    res <- limmaRes$res
    results <- limmaRes$results
    
    ## FREQUENCY MATRIX: Reading intermediate files to produce a frequency
    ## matrix
    message(.timestamp(), " -- Summarizing results...")
    intermediate <- .freqMatrix(limmaRes, data, n1, ifelse(unsup, 0, n2), 
        r, multi, unsup)
    
    # Matrix of combinations and number of events.
    ord <- order(results[, "nFeatures"], decreasing = TRUE)
    res$results <- results[ord, -(r * 2 + 2)]
    res$pos.iter <- length(limmaRes$limma1)/2
    
    if (!unsup & !multi) {
        res$incidenceMatrix <- gdata::interleave(intermediate$UP, intermediate$DOWN, 
            append.source = TRUE, sep = "deco", drop = FALSE)
    } else {
        res$incidenceMatrix <- intermediate$MULTI
    }
    
    if (length(intermediate$both.feat) > 0) {
        res$incidenceMatrix <- rbind(res$incidenceMatrix, intermediate$both.feat.mx)
    }
    
    res$incidenceMatrix <- res$incidenceMatrix[, colnames(res$incidenceMatrix) %in% 
        colnames(data)]
    
    ### CALCULATING FINAL STATISTICS Creating a table to contain all statistical
    ### information about features.
    j <- length(limmaRes$limma1)/2
    top_eje <- intermediate$top_eje[2:length(rownames(intermediate$top_eje)), 
        ]
    res$subStatFeature <- as.data.frame(matrix(ncol = 12, nrow = length(unique(top_eje[, 
        c("ID")]))))
    colnames(res$subStatFeature) <- c("ID", "UpDw", "Avrg.logFC", "Best.adj.P.Val", 
        "Repeats", "FR.Repeats", "RF.Positive.Repeats", "Chi.Square", "P.Val.ChiSq", 
        "ChiSq.adj.P.Val.FDR")
    res$subStatFeature[, c("ID")] <- unique(top_eje[, c("ID")])
    res$subStatFeature[, 3:10] <- 0
    ncomb <- dim(results)[1]
    
    # ## Summarizing statistics...
    message(.timestamp(), " -- Calculating statistics...")
    suppressWarnings(limma2 <- bplapply(seq_len(dim(res$subStatFeature)[1]), 
        FUN = .statCalc, BPPARAM = bpparam, tab = res$subStatFeature, top_eje, 
        ncomb, j))
    
    res$subStatFeature <- as.data.frame(do.call(rbind, limma2))
    res$subStatFeature[, 3:10] <- suppressWarnings(apply(res$subStatFeature[, 
        3:10], 2, as.numeric))
    
    message(.timestamp(), " -- Computed ", length(unique(top_eje[, c("ID")])), 
        " features.\n")
    
    ## Adjusting p.values from Chi.Square.
    res$subStatFeature[, c("ChiSq.adj.P.Val.FDR")] <- p.adjust(res$subStatFeature[, 
        c("P.Val.ChiSq")], method = "fdr", n = length(res$subStatFeature[, 
        c("P.Val.ChiSq")]))
    
    # Ordering columns...
    res$subStatFeature <- res$subStatFeature[, c("ID", "UpDw", "Avrg.logFC", 
        "Best.adj.P.Val", "Repeats", "FR.Repeats", "RF.Positive.Repeats", 
        "Chi.Square", "P.Val.ChiSq", "ChiSq.adj.P.Val.FDR")]
    
    ## Annotating features using 'AnnotateDECO' function.
    if (annot) {
        infogenes <- AnnotateDECO(ids = as.character(res$subStatFeature[, 
            c("ID")]), id.type = id.type, attributes = attributes, pack.db = pack.db)
        res$subStatFeature <- cbind(res$subStatFeature, infogenes[as.vector(res$subStatFeature[, 
            c("ID")]), ])
    }
    # Making up all statistics.
    res$subStatFeature <- droplevels.data.frame(as.data.frame(res$subStatFeature))
    res$subStatFeature[, 3:10] <- suppressWarnings(apply(res$subStatFeature[, 
        3:10], 2, as.numeric))
    if (!unsup & !multi) {
        res$subStatFeature[, c("UpDw")] <- intermediate$UpDw[match(as.character(res$subStatFeature[, 
            c("ID")]), names(intermediate$UpDw))]
    } else {
        res$subStatFeature <- res$subStatFeature[, colnames(res$subStatFeature) != 
            "UpDw"]
    }
    
    # Standard.Chi.Square
    res$subStatFeature[, "Standard.Chi.Square"] <- ((as.numeric(res$subStatFeature[, 
        c("Chi.Square")]) - (2 * as.numeric(res$subStatFeature[, c("Repeats")])))/sqrt(4 * 
        as.numeric(res$subStatFeature[, c("Repeats")])))
    
    res$subStatFeature <- res$subStatFeature[order(res$subStatFeature$Standard.Chi.Square, 
        decreasing = TRUE), ]
    rownames(res$subStatFeature) <- res$subStatFeature[, "ID"] <- as.character(res$subStatFeature[, 
        "ID"])
    
    # Removing temporary dir and all intermediate files generated.
    unlink(temp.path, recursive = TRUE)
    
    message(.timestamp(), " -- Done.\n")
    
    return(res)
}


################################# decoNSCA ### Function to carry out factorial analysis with NSCA
################################# (Non-Symmetrical Correspondence Analysis) of the incidenceMatrix
################################# generated by 'DECO' subsampling function.

decoNSCA <- function(sub, v = 80, k.control = NULL, k.case = NULL, rep.thr = 3, 
    bpparam = SerialParam(), samp.perc = 0.05, method = "ward.D") {
    call <- match.call()
    
    # Agglomeration method for hierarchical clustering of samples.
    if (all(!(is.null(method))) && !(method %in% c("ward.D", "ward.D2", "single", 
        "complete", "average", "mcquitty", "median", "centroid"))) {
        stop("ERROR: Input 'method.heatmap' have to be one of 'hclust'{stats} function methods:
         ward.D, ward.D2, single, complete, average, mcquitty, median or centroid")
    }
    
    data <- sub$data
    
    # Variable with original IDs used to substitute modified IDs of
    # incidenceMatrix.
    message(.timestamp(), " -- Applying repeats threshold...\n")
    
    ## Threshold for Repeats
    Sub <- .repThr(sub, rep.thr, samp.perc)
    g.names <- Sub$g.names
    sub <- Sub$sub
    
    # Setting up both classes and control for 'binary' design.
    if (all(!(is.na(sub$classes))) & length(levels(sub$classes)) <= 2) {
        # Instructions just for 'binary' RDA design.
        if (is.na(sub$control)) {
            cl1 <- levels(sub$classes)[1]
            cl2 <- levels(sub$classes)[2]
            ind1 <- which(sub$classes == cl1)
            ind2 <- which(sub$classes == cl2)
        } else {
            cl1 <- sub$control
            cl2 <- levels(sub$classes)[which(levels(sub$classes) != sub$control)]
            ind1 <- which(sub$classes == cl1)
            ind2 <- which(sub$classes == cl2)
            sub$classes <- sub$classes[c(ind1, ind2)]
        }
        n1 <- length(ind1)
        n2 <- length(ind2)
        sub$incidenceMatrix <- sub$incidenceMatrix[, names(sub$classes)]
        
        # Calculating raw 'delta.signal' differences between classes for all
        # differential features.
        d <- sub$data[rownames(sub$data) %in% rownames(sub$subStatFeature), 
            ]
        
        delta.signal <- rowMeans(d[, (n1 + 1):(n1 + n2)]) - rowMeans(d[, seq_len(n1)])
        
        sub$subStatFeature <- data.frame(sub$subStatFeature[order(sub$subStatFeature[, 
            c("ID")]), ], delta.signal = delta.signal[order(names(delta.signal))])
        # Standard deviation per class
        sd.Ctrl <- apply(sub$data[rownames(sub$subStatFeature), seq_len(n1)], 
            1, sd)
        sd.Case <- apply(sub$data[rownames(sub$subStatFeature), (n1 + 1):(n1 + 
            n2)], 1, sd)
        
        # Classification of feature profiles: 'ideal', 'generic', 'specific, and
        # 'both'.
        overlap <- overlapC(sub, cl1, bpparam)
        
        profile <- overlap$profile
        overlap <- overlap$overlap
        
        UpDw <- sub$subStatFeature$UpDw
        names(UpDw) <- as.character(sub$subStatFeature$ID)
    } else {
        n1 <- dim(data)[2]
        n2 <- dim(data)[2]
        SD <- apply(sub$data, 1, function(x) sd(x, na.rm = TRUE))
    }
    # Applying NSCA to all samples or both classes.
    z <- 1
    while (z %in% seq_len(2)) {
        # NSCA for 'unsupervised' and 'multiclass' design.
        if (all(is.na(sub$classes)) || length(levels(sub$classes)) > 2) {
            # Removing features or samples with any differential event.
            mx <- sub$incidenceMatrix[rowSums(sub$incidenceMatrix) > 0, ]
            mx <- mx[, colSums(sub$incidenceMatrix) > 0]
            message(.timestamp(), "-- Calculating all samples together...")
            # NSCA
            nsc.res <- NSCACluster(mx = mx, data = sub$data, k = k.control, 
                method.dend = method, id.names = g.names, v = v)
            k <- k.control
            diffMX <- cbind(sub$subStatFeature, nsc.res$infoFeature[rownames(sub$subStatFeature), 
                ])
            diffMX <- cbind(diffMX, sd = SD[rownames(sub$subStatFeature)])
            # Order for columns of final table with statistical information.
            coln <- c("ID", "SYMBOL", "hgnc_symbol", "Repeats", "Repeats.index", 
                "FR.Repeats", "Avrg.logFC", "Standard.Chi.Square", "P.Val.ChiSq", 
                "ChiSq.adj.P.Val.FDR", "sd", "Tau.feature", "Dendrogram.group", 
                "h.Best", "h.Range", "GENENAME")
            ord <- na.omit(match(c(coln, colnames(diffMX)[!colnames(diffMX) %in% 
                coln]), colnames(diffMX)))
            z <- 3
        }
        # NSCA for 'control' samples
        if (z == 1) {
            # Removing features or samples with any differential event.
            mx <- sub$incidenceMatrix[rowSums(sub$incidenceMatrix[, seq_len(n1)]) > 
                0, seq_len(n1)]
            mx <- mx[, colSums(sub$incidenceMatrix[, seq_len(n1)]) > 0]
            message(.timestamp(), "-- Calculating control group...")
            # NSCA
            nsc.res1 <- NSCACluster(mx = mx, data = sub$data, k = k.control, 
                method.dend = method, UpDw = UpDw, dir = "UP", id.names = g.names, 
                v = v, label = cl1, raw = rowMeans(sub$data[, colnames(mx)]))
            colnames(nsc.res1$infoFeature) <- paste(colnames(nsc.res1$infoFeature), 
                "Ctrl", sep = ".")
            diffMX <- cbind(sub$subStatFeature, sd.Ctrl = sd.Ctrl[rownames(sub$subStatFeature)], 
                nsc.res1$infoFeature[rownames(sub$subStatFeature), c("Tau.feature.Ctrl", 
                  "Dendrogram.group.Ctrl", "h.Best.Ctrl", "h.Range.Ctrl")])
        }
        # NSCA for 'case' samples
        if (z == 2) {
            # Removing features or samples with any differential event.
            mx <- sub$incidenceMatrix[rowSums(sub$incidenceMatrix[, (n1 + 
                1):(n1 + n2)]) > 0, (n1 + 1):(n1 + n2)]
            mx <- mx[, colSums(sub$incidenceMatrix[, (n1 + 1):(n1 + n2)]) > 
                0]
            message(.timestamp(), " -- Calculating case group...")
            # NSCA
            nsc.res2 <- NSCACluster(mx = mx, data = sub$data, k = k.case, 
                method.dend = method, UpDw = UpDw, dir = "DOWN", id.names = g.names, 
                v = v, label = cl2, raw = rowMeans(sub$data[, colnames(mx)]))
            colnames(nsc.res2$infoFeature)[seq_len(length(colnames(nsc.res2$infoFeature)) - 
                1)] <- paste(colnames(nsc.res2$infoFeature)[seq_len(length(colnames(nsc.res2$infoFeature)) - 
                1)], "Case", sep = ".")
            diffMX <- cbind(diffMX, sd.Case = sd.Case[rownames(sub$subStatFeature)], 
                overlap.Ctrl.Case = overlap[rownames(sub$subStatFeature)], 
                nsc.res2$infoFeature[rownames(sub$subStatFeature), c("Tau.feature.Case", 
                  "Dendrogram.group.Case", "h.Best.Case", "h.Range.Case")])
            # Standard.Chi.Square calculation.
            diffMX <- cbind(diffMX, Profile = profile[rownames(sub$subStatFeature)])
            # Order for columns of final table with statistical information.
            coln <- c("ID", "SYMBOL", "hgnc_symbol", "UpDw", "Profile", "overlap.Ctrl.Case", 
                "Repeats", "Repeats.index", "FR.Repeats", "delta.signal", 
                "Avrg.logFC", "Standard.Chi.Square", "P.Val.ChiSq", "ChiSq.adj.P.Val.FDR", 
                "sd.Ctrl", "Tau.feature.Ctrl", "Dendrogram.group.Ctrl", "h.Best.Ctrl", 
                "h.Range.Ctrl", "sd.Case", "Tau.feature.Case", "Dendrogram.group.Case", 
                "h.Best.Case", "h.Range.Case", "GENENAME")
            ord <- na.omit(match(c(coln, colnames(diffMX)[!colnames(diffMX) %in% 
                coln]), colnames(diffMX)))
        }
        z <- z + 1
    }
    # Making up of final statistical table.
    diffg <- data.frame(diffMX[, c(ord)], diffMX[, na.action(ord)])
    colnames(diffg) <- c(colnames(diffMX)[c(ord)], colnames(diffMX)[na.action(ord)])
    # Creating final result object.
    if (!(all(is.na(sub$classes))) & !length(levels(sub$classes)) > 2) {
        ord <- with(diffg, order(Profile, -Standard.Chi.Square))
        diffg <- diffg[ord, ]
        results <- new("deco", featureTable = as.data.frame(diffg[, !duplicated(colnames(diffg))]), 
            NSCAcluster = list(Control = nsc.res1, Case = nsc.res2), incidenceMatrix = sub$incidenceMatrix, 
            classes = as.factor(sub$classes), pos.iter = sub$pos.iter, control = as.character(sub$control), 
            q.val = sub$q.val, rep.thr = rep.thr, samp.perc = samp.perc, subsampling.call = sub$call, 
            nsca.call = call)
    } else {
        ord <- order(rowMeans(apply(cbind(diffg$Standard.Chi.Square, diffg$Standard.Chi.Square, 
            diffg$Repeats, diffg$sd, diffg$h.Range), 2, function(x) rank(-x))))
        diffg <- diffg[ord, ]
        results <- new("deco", featureTable = as.data.frame(diffg[, !duplicated(colnames(diffg))]), 
            NSCAcluster = list(All = nsc.res), incidenceMatrix = sub$incidenceMatrix, 
            classes = as.factor(sub$classes), pos.iter = sub$pos.iter, control = as.character(sub$control), 
            q.val = sub$q.val, rep.thr = rep.thr, samp.perc = samp.perc, subsampling.call = sub$call, 
            nsca.call = call)
    }
    
    return(results)
}
